L74-java-channel-creds.md

```diff
@@ -0,0 +1,489 @@
+    public static Result negotiator(ProtocolNegotiator.Factory factory) {...}
+    public Result withCallCredentials(CallCredentials callCreds) {...}
+  }
+}
+```
+
+### ServerCredentials
+
+Server-side will be a clone of client-side, but using the `Server` instead of
+`Channel`. `ServerCredentials` cannot contain `CallCredentials`, so
+`CompositeServerCredentials` will be omitted. There is also currently no known
+use-case for `ChoiceServerCredentials`, so it will be omitted.
+
+```java
+// instead of:
+server = ServerBuilder.forPort(443)
+    .useTransportSecurity(new File("cert.pem"), new File("cert.key"))
+    .build().start();
+server = ServerBuilder.forPort(8080)
+    .build().start();
+// the user would now:
+ChannelCredentials tlsCreds = new TlsServerCredentials(
+    new File("cert.pem"), new File("cert.key"));
+server = Grpc.newServerBuilderForPort(443, tlsCreds)
+    .build().start();
+server = Grpc.newServerBuilderForPort(8080, new InsecureChannelCredentials())
+    .build().start();
+```
+
+
+```java
+package io.grpc;
+
+public abstract class ServerCredentials {}
+```
+
+Since servers using TLS require a certificate, the `TlsServerCredentials`
+cannot have the simple no-arg configuration of `TlsChannelCredentials`. It will
+provide convenience constructors for providing a certificate and unencrypted
+key. If any other configuration is necessary, the `Builder` would be used.
+
+```java
+package io.grpc;
+
+public final class TlsServerCredentials extends ServerCredentials {
+  /**
+   * Creates an instance using provided certificate chain and private key.
+   * Generally they should be PEM-encoded and the key is an unencrypted PKCS#8
+   * key (file headers have "BEGIN CERTIFICATE" and "BEGIN PRIVATE KEY").
+   */
+  public TlsServerCredentials(File certChain, File privateKey)
+      throws IOException {...}
+  // Will not auto-close InputStream, unlike existing transportSecurity()
+  public TlsServerCredentials(InputStream certChain, InputStream privateKey)
+      throws IOException {...}
+
+  /** Same as client-side */
+  public Set<Feature> incomprehensible(Set<Feature> understoodFeatures) {...}
+
+  public byte[] getCertificateChain() { return Arrays.copyOf(...); }
+  public byte[] getPrivateKey() { return Arrays.copyOf(...); }
+  public String getPassword() {...}
+
+  /* Other methods can be added to support various Features */
+
+  public enum Feature {}
+
+  public static Builder newBuilder() { return new Builder(); }
+
+  public final static class Builder {
+    public Builder keyManager(File certChain, File privateKey)
+        throws IOException {...}
+    public Builder keyManager(File certChain, File privateKey, String password)
+        throws IOException {...}
+    public Builder keyManager(InputStream certChain, InputStream privateKey)
+        throws IOException {...}
+    public Builder keyManager(
+        InputStream certChain, InputStream privateKey, String password)
+        throws IOException {...}
```

Comment:
We'll also need the equivalent of netty's `ClientAuth` for making client authentication optional vs required when root-certs are provided.

Reply:
That will be part of a client authentication Feature of some sort. You need trust roots and ASK/REQUIRE. My goal here is to reproduce what we already have in the new API and have new features be introduced afterward.

Reply:
I only added keyPassword support now because it is relatively easy to handle and is closely associated with what we already support; it'd be more work for implementations to deal with a Feature than just have it be assumed initially.

Reply:
ok